lib/ClangImporter/ClangImporter.cpp

```diff
@@ -4044,3 +4044,22 @@ swift::getModuleCachePathFromClang(const clang::CompilerInstance &Clang) {
   return llvm::sys::path::parent_path(SpecificModuleCachePath).str();
 }
 
+clang::FunctionDecl *ClangImporter::generateCXXFunctionTemplateSpecializations(
+    ASTContext &ctx, clang::FunctionTemplateDecl *func, SubstitutionMap subst) {
+  SmallVector<clang::TemplateArgument, 4> templateSubst;
+  if (!ctx.getClangTemplateArguments(func->getTemplateParameters(),
+                                     subst.getReplacementTypes(),
+                                     templateSubst)) {
+    llvm::report_fatal_error(
```

Comment:
@varungandhi-apple is this the right way to handle errors?

Reply:
As a stop-gap solution, yes (if you're taking this route, please add a `FIXME:` and file a JIRA, assigning it to who will handle it). As a long-term solution, no. Imagine trying to call a templated C++ function and the compiler crashed on you. Not fun.

This function's return type would probably need to be `llvm::Expected<clang::FunctionDecl *>`. In the error case, you'd return some information about which generic arguments failed to convert (ideally, if multiple arguments failed to convert, I think you should return all the arguments which failed to convert). I suspect this will require a new `class TemplateInstantiationError : public llvm::ErrorInfo<TemplateInstantiationError>` to wrap the state for the error case.

Looks like Clang emits its own diagnostics separately in case there is an instantiation error, so you probably don't need to do any handling for that here. Please add some basic tests to make sure that we are surfacing Clang template errors.

I'm not entirely sure at what point in the call-stack we should emit the diagnostic. @CodaFi -- do you have suggestions here?

Reply:
I'd prefer to implement this with proper error handling. Otherwise, I have a feeling it might not get done (at least for a while). 

I can work on that tomorrow. 

What's the best way to tell the compiler it needs to fail and exit?

Reply:
Generally, the steps are:

1. Create a new diagnostic in `DiagnosticSema.def` (and also in `localization/diagnostics/en.yaml`, there is some temporary duplication). The diagnostic is defined in a little DSL using macros. You will need to define a new `ERROR` diagnostic. I suspect we don't support variadic formatting for diagnostics, so you might need to additionally string-ify things manually before passing a `StringRef` for the types etc. to the diagnostic.
2. Grab hold of the ASTContext where you want to emit the diagnostic and do `ctx.Diags.diagnose(...)` and emit the diagnostic. There is a `SuppressDiagnostics` setting on `ExprRewriter` (couple of functions up the call stack) which might need to be checked before emitting the diagnostic, I'm not 100% sure when it should/should not be checked.

Reply:
I've added the diagnostics and a test to test them. Let me know what you think. 

I originally implemented this using `llvm::Expected` but that ended up being a lot of boilerplate that wasn't really useful and it didn't seem to give me any benefit. If you want, I can reimplement that version, though.

Reply:
Thank you for the changes. This approach is also good. 👍🏼

I think we should rephrase the diagnostics to match our existing diagnostic style better; at the moment, the description is in terms of the implementation, not in terms of a user's model of what is happening. We can change that in a follow-up PR though, it's not super urgent.

Reply:
OK. I'll add a TODO comment. For the next ~year, the implementers are probably going to be the only people seeing this error, though, so for now, I think this error message might actually be more helpful.